wagtail/tests/test_page_model.py

```diff
@@ -210,6 +212,36 @@ def setUp(self):
         self.unrecognised_port = "8000"
         self.unrecognised_hostname = "unknown.site.com"
 
+    def test_route_for_request_query_count(self):
+        request = get_dummy_request(site=self.events_site)
+        with self.assertNumQueries(2):
+            # expect queries for site & page
+            Page.route_for_request(request, request.path)
+        with self.assertNumQueries(0):
+            # subsequent lookups should be cached on the request
+            Page.route_for_request(request, request.path)
+    
+    def test_route_for_request_value(self):
+        request = get_dummy_request(site=self.events_site)
+        self.assertFalse(hasattr(request, '_wagtail_route_for_request'))
+        result = Page.route_for_request(request, request.path)
+        self.assertTrue(isinstance(result, RouteResult))
+        self.assertEqual((result[0], result[1], result[2]), (self.events_site.root_page.specific, [], {}))
+        self.assertTrue(hasattr(request, '_wagtail_route_for_request'))
+        self.assertIs(request._wagtail_route_for_request, result)
+    
+    def test_route_for_request_cached(self):
+        request = get_dummy_request(site=self.events_site)
+        m = Mock()
+        request._wagtail_route_for_request = m
+        with self.assertNumQueries(0):
+            self.assertEqual(Page.route_for_request(request, request.path), m)
+    
+    def test_route_for_request_404(self):
+        request = get_dummy_request(path='does-not-exist', site=self.events_site)
+        with self.assertRaises(Http404):
```

Comment:
Nit: I feel like a generic API raising `Http404` ties it a little too closely to the view logic for me. Perhaps this should be caught and return `None` instead?

Reply:
I think `route_for_request` should avoid it too. `Http404` only really makes sense in the request-response lifecycle.

Reply:
I guess I am confused because this is a request-response lifecycle method - everywhere that it is used needs to raise a 404 if page is None. [The logic is moved as-is from the serve view so that middleware can use it before the view is processed.](https://github.com/wagtail/wagtail/pull/11683/files#diff-8e14e2bcb91ddb809d596250d4c244bf90f41480d346129ce6b98532cb58d490L12) But I will submit a follow up commit to suppress the 404 and get this moving shortly.